crates/approx_derive/src/lib.rs

```diff
@@ -0,0 +1,136 @@
+use proc_macro2::TokenStream;
+use proc_macro_error::{abort, proc_macro_error, OptionExt, ResultExt};
+use quote::quote;
+use syn::{
+    parse_macro_input,
+    punctuated::{self},
+    Attribute, Data, DeriveInput, Lit, Meta, MetaNameValue, NestedMeta, Type,
+};
+
+#[proc_macro_derive(AbsDiffEq, attributes(abs_diff_eq))]
+#[proc_macro_error]
+pub fn abs_diff_eq(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    generate_abs_diff_eq(input).into()
+}
+
+fn generate_abs_diff_eq(input: DeriveInput) -> TokenStream {
+    let fields = match input.data {
+        Data::Struct(data) => data.fields,
+        Data::Enum(data) => abort!(
+            data.enum_token,
+            "`AbsDiffEq` can only be derived for `struct`",
+        ),
+        Data::Union(data) => abort!(
+            data.union_token,
+            "`AbsDiffEq` can only be derived for `struct`",
+        ),
+    };
+    let epsilon = extract_epsilon(&input.attrs).expect_or_abort("`epsilon` not specified");
+    let name = input.ident;
+
+    let conditions = fields.into_iter().map(|field| {
+        let identifier = field
+            .ident
+            .clone()
+            .unwrap_or_else(|| abort!(field, "field has to be named"));
+        quote! {
+            self.#identifier.abs_diff_eq(&other.#identifier, epsilon)
+        }
+    });
+
+    quote! {
+        impl approx::AbsDiffEq for #name {
+            type Epsilon = #epsilon;
+
+            fn default_epsilon() -> Self::Epsilon {
+                Self::Epsilon::default_epsilon()
+            }
+
+            fn abs_diff_eq(&self, other: &Self, epsilon: Self::Epsilon) -> bool {
+                #(#conditions)&&*
+            }
+        }
+    }
+}
+
+fn extract_epsilon(attrs: &[Attribute]) -> Option<Type> {
+    attrs
+        .iter()
+        .filter_map(parse_meta_items)
+        .flatten()
+        .find_map(|meta| match meta {
+            NestedMeta::Meta(Meta::NameValue(MetaNameValue { path, lit, .. }))
+                if path.is_ident("epsilon") =>
+            {
+                let string = match lit {
+                    Lit::Str(string) => string,
+                    _ => abort!(lit, "expected string literal"),
+                };
+                let epsilon = string
+                    .parse()
+                    .expect_or_abort("failed to parse epsilon type");
+                Some(epsilon)
+            }
+            _ => None,
+        })
+}
+
+fn parse_meta_items(attribute: &Attribute) -> Option<punctuated::IntoIter<NestedMeta>> {
```

Comment:
This parses just one item, doesn't it?
```suggestion
fn parse_meta_item(attribute: &Attribute) -> Option<punctuated::IntoIter<NestedMeta>> {
```

Reply:
hmmm... :thinking: there is one `Meta`, but several items in that meta... And the function is returning an iterator (-of multiple items)